### ISSUE TITLE
Conflict with env function

### ISSUE THREAD

**Issue by oscarotero** [cakephp/phinx#1647]
Hi.
My site stopped working because I have a `env` function to return environment variables and I just realized this library use CakePHP, that defines another `env` function before, so the site was using the cake's function that have a different behaviour.
Is there any way to have phinx installed without having these functions defined by default? (Maybe using a phar file or something like that).

Thanks!

**Comment by dereuromark** [comment]
Shouldnt this redirect to the new docs then? @markstory

**Comment by markstory** [comment]
>  Is there any way to have phinx installed without having these functions defined by default? (Maybe using a phar file or something like that).

We don't currently create phar files. I'm not too familiar with the tooling to do so either, perhaps another contributor is.

**Comment by dereuromark** [comment]
The same trick might also work here: https://github.com/cakephp/phinx/issues/1401#issuecomment-435210854

**Comment by oscarotero** [comment]
> The same trick might also work here: #1401 (comment)

Yes and no 

I'm using `oscarotero/env` library to get environment variables and it provides a `env()` function (https://github.com/oscarotero/env#the-env-function)

The library does not expose that function by default (it is not autoloaded by composer), you have to run `Env::init()` so it's not posible to execute this funcion before include composer autoload php file. 

The only workaround is to don't use the `env` function (but `Env::get()` instead). Other option is include manually the function before include composer (https://github.com/oscarotero/env/blob/master/src/env_function.php).

Anyway, I've started to using this library before it was converted to a cake component and would like it to remain a standalone framework-agnostic tool, so IMO, polluting the global with these functions is not a good idea. Global functions should be defined/loaded and used by final users, but not by dependencies. Maybe cake could expose its funcions under its own namespace, or using a class with static methods.

**Comment by dereuromark** [comment]
@markstory Any way we can do this?

**Comment by markstory** [comment]
The env() function in Cake is conditionally defined. If your application defines the function first the Cake one won't be created.

Not autoloading the env() function entirely could be possible, but runs a high risk of breaking compatibility unless we can come up with a way to ensure env() exists everywhere but in phinx's usage.

**Comment by dereuromark** [comment]
We should add a docs example how to load custom one before the core.
That would solve it for n ow.

**Comment by TimurFlush** [comment]
Pleasee see https://github.com/TimurFlush/phinx2phar

**Comment by rquadling** [comment]
A library that has some helper functions but pollutes the global namespace is really not very nice. Move the global methods to the Cake specific namespace, and heavily document a polyfiller project that reinstates the global pollution with a strong suggestion that the consumer's code should be upgraded to use the Cake specific namespace if they want to continue using the global function.

I'm unsure if we should be using a static class, as there is a small semantic difference between a static method and a namespaced class (so small I can hardly see it).

This is one of my biggest gripes with another well known and popular framework that starts with L.

In short why does ANY framework think it should be able to pollute the global namespace.

**Comment by lorenzo** [comment]
I just with PHP had function autoloading

**Comment by rquadling** [comment]
https://wiki.php.net/rfc/function_autoloading

**Comment by pooriamo** [comment]
I use psysch REPL library which includes `autoload.php` in runtime. So I can't define my own `env` function before that.
